ecl/hql/hqlfold.cpp

```diff
@@ -1335,8 +1346,486 @@ IValue * foldExternalCall(IHqlExpression* expr, unsigned foldOptions, ITemplateC
+                    size = *(size32_t *) data;
+                    data += sizeof(size32_t);
+                }
+                values.append(*createConstant(createDataValue(data, size)));
+                break;
+            case type_unicode:
+                if (size==UNKNOWN_LENGTH)
+                {
+                    numChars = *(size32_t *) data;  // in characters
+                    data += sizeof(size32_t);
+                    values.append(*createConstant(createUnicodeValue((const UChar *) data, numChars, LINK(childType))));
+                    size = numChars * sizeof(UChar);
+                }
+                else
+                {
+                    values.append(*createConstant(createUnicodeValue((const UChar *) data, numChars, LINK(childType))));
+                }
+                break;
+            case type_utf8:
+                // size is always UNKNOWN_LENGTH for uft8
+                assertex(size==UNKNOWN_LENGTH);
+                numChars = *(size32_t *) data;  // in characters
+                data += sizeof(size32_t);
+                values.append(*createConstant(createUtf8Value(numChars, data, LINK(childType))));
+                size = rtlUtf8Size(numChars, data);
+                break;
+            default:
+                return NULL;
+            }
+            if (size != UNKNOWN_LENGTH)
+                data += size;
+        }
+        return createValue(no_list, LINK(type), values);
+    }
+}
+
+IHqlExpression * foldEmbeddedCall(IHqlExpression* expr, unsigned foldOptions, ITemplateContext *templateContext)
+{
+    if (!checkEmbeddedFoldable(expr, foldOptions))
+        return NULL;
+    IHqlExpression * funcdef = expr->queryBody()->queryFunctionDefinition();
+    IHqlExpression * outofline = funcdef->queryChild(0);
+    assertex(outofline->getOperator() == no_outofline);
+    IHqlExpression * body = outofline->queryChild(0);
+    assertex(body->getOperator()==no_embedbody);
+    ITypeInfo * returnType = funcdef->queryType()->queryChildType();
+
+    bool isImport = body->hasAttribute(importAtom);
+    IHqlExpression * formals = funcdef->queryChild(1);
+    assertex(formals->numChildren() == expr->numChildren());  // MORE - do default params change this?
+    unsigned flags = isImport ? EFimport : EFembed;
+    if (formals->numChildren()==0)
+        flags |= EFnoparams;
+    if (returnType->getTypeCode()==type_void)
+        flags |= EFnoreturn;
+
+    StringBuffer optionsStr;
+    ForEachChild(idx, body)
+    {
+        IHqlExpression *child = body->queryChild(idx);
+        if (child->isAttribute() && !isInternalEmbedAttr(child->queryName()))
+        {
+            if (optionsStr.length())
+                optionsStr.append(",");
+            optionsStr.append(child->queryName());
+            IHqlExpression * value = child->queryChild(0);
+            if (value)
+            {
+                optionsStr.append("=");
+                value->queryValue()->getUTF8Value(optionsStr);
```

Comment:
I put an isConstant() check in checkExternalFoldable - but it's always returning false...